### PR TITLE
Implement smithing table

### PR DESCRIPTION
**Pull request by joserobjr**
In survival mode, the client is sending inventory mismatch when attempting to get a netherite pickaxe and I have no idea why yet.

### CODE REVIEW

**Comment by PocketMiner82** [comment]
@joserobjr Maybe this will help you: https://github.com/GeyserMC/Geyser/pull/1360

**Comment by Camotoy** [comment]
To my understanding, without switching to the new inventory system you will get INVENTORY_MISMATCH when attempting to take an item out of the smithing table's output.

**Comment by joserobjr** [comment]
I think I'm going to do the same hack as geyser

**Comment by Camotoy** [comment]
Geyser is not implementing this hack - instead, we're working on implementing the new inventory system that allows the smithing table to work.

File: src/main/java/cn/nukkit/inventory/CraftingManager.java [hunk 1]
```diff
@@ -95,6 +103,25 @@ public CraftingManager() {
 
         log.info("Loaded {} recipes.", this.recipes.size());
     }
+    
+    private void registerSmithingRecipes() {
```
Comment (LoboMetalurgico) [inline src/main/java/cn/nukkit/inventory/CraftingManager.java:107]: Missing @PowerNukkitOnly

File: src/main/java/cn/nukkit/inventory/CraftingManager.java [hunk 2]
```diff
@@ -554,6 +656,15 @@ public CraftingRecipe matchRecipe(List<Item> inputList, Item primaryOutput, List
         return null;
     }
 
+    private boolean matchItemsAccumulation(SmithingRecipe recipe, List<Item> inputList, Item primaryOutput) {
```
Comment (LoboMetalurgico) [inline src/main/java/cn/nukkit/inventory/CraftingManager.java:665]: Missing @PowerNukkitOnly or @PowerNukkitDifference

File: src/main/java/cn/nukkit/inventory/InventoryType.java [hunk 3]
```diff
@@ -28,23 +28,30 @@
...
     CAMPFIRE(4, "Campfire", -9), // -9 = NONE
     @PowerNukkitOnly @Since("1.4.0.0-PN") ENTITY_EQUIPMENT(36, "Entity Equipment", -1), //36 CONTAINER
     @PowerNukkitOnly @Since("1.4.0.0-PN") ENTITY_ARMOR(4, "Entity Armor", -1), //4 ARMOR
-    MINECART_CHEST(27, "Minecart with Chest", 0),
-    MINECART_HOPPER(5, "Minecart with Hopper", 8),
-    OFFHAND(1, "Offhand", -1);
+    MINECART_CHEST(27, "Minecart with Chest", 0), // Should be 10
+    MINECART_HOPPER(5, "Minecart with Hopper", 8), // Should be 11
+    OFFHAND(1, "Offhand", -1),
+    SMITHING_TABLE(2, "Smithing Table", 33);
```
Comment (LoboMetalurgico) [inline src/main/java/cn/nukkit/inventory/InventoryType.java:54]: Missing @PowerNukkitOnly

File: src/main/java/cn/nukkit/inventory/CraftingManager.java [hunk 4]
```diff
@@ -55,6 +59,8 @@
     private static int RECIPE_COUNT = 0;
     protected final Map<Integer, Map<UUID, ShapelessRecipe>> shapelessRecipes = new Int2ObjectOpenHashMap<>();
     protected final Map<Integer, Map<UUID, CartographyRecipe>> cartographyRecipes = new Int2ObjectOpenHashMap<>();
+    
+    private final Int2ObjectOpenHashMap<Map<UUID, SmithingRecipe>> smithingRecipes = new Int2ObjectOpenHashMap<>();
```
Comment (LoboMetalurgico) [inline src/main/java/cn/nukkit/inventory/CraftingManager.java:63]: Missing @PowerNukkitOnly

File: src/main/java/cn/nukkit/inventory/SmithingInventory.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+    @PowerNukkitOnly
+    @Since("1.4.0.0-PN")
+    @Nullable
+    public SmithingRecipe matchRecipe() {
+        return Server.getInstance().getCraftingManager().matchSmithingRecipe(getEquipment(), getIngredient());
+    }
+
+    @Override
+    public void onSlotChange(int index, Item before, boolean send) {
```
Comment (LoboMetalurgico) [inline src/main/java/cn/nukkit/inventory/SmithingInventory.java:63]: Missing @PowerNukkitOnly

File: src/main/java/cn/nukkit/inventory/SmithingInventory.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        SmithingRecipe recipe = matchRecipe();
+        if (recipe == null) {
+            result =  Item.get(0);
+        } else {
+            result = recipe.getFinalResult(getEquipment());
+        }
+        setResult(result);
+    }
+    
+    private void setResult(Item result) {
```
Comment (LoboMetalurgico) [inline src/main/java/cn/nukkit/inventory/SmithingInventory.java:83]: Missing @PowerNukkitOnly

File: src/main/java/cn/nukkit/inventory/transaction/action/InventoryAction.java [hunk 7]
```diff
@@ -3,10 +3,12 @@
 import cn.nukkit.Player;
 import cn.nukkit.inventory.transaction.InventoryTransaction;
 import cn.nukkit.item.Item;
+import lombok.ToString;
 
 /**
  * @author CreeperFace
  */
+@ToString
```
Comment (LoboMetalurgico) [inline src/main/java/cn/nukkit/inventory/transaction/action/InventoryAction.java:11]: Need to be @ToString(callSuper = true)

**Review by LoboMetalurgico — CHANGES_REQUESTED** [review]
Helo! Thank you for your Pull Request. It is really very good and contains only a few annotation problems. Could you fix them?

**Review by LoboMetalurgico — APPROVED** [review]
Thank You!

**Pull request merged**